electrum/plugins/jade/jade.py

```diff
@@ -0,0 +1,471 @@
+from electrum.transaction import Transaction
+from electrum.wallet import Multisig_Wallet
+from electrum.util import UserFacingException
+from electrum.base_wizard import ScriptTypeNotSupported
+from electrum.logging import get_logger
+from electrum.plugin import runs_in_hwd_thread, Device
+from electrum.network import Network
+
+from ..hw_wallet import HW_PluginBase, HardwareClientBase
+from ..hw_wallet.plugin import OutdatedHwFirmwareException
+
+
+_logger = get_logger(__name__)
+
+#import logging
+#LOGGING = logging.INFO
+#if LOGGING:
+#    logger = logging.getLogger('jade')
+#    logger.setLevel(LOGGING)
+#    device_logger = logging.getLogger('jade-device')
+#    device_logger.setLevel(LOGGING)
+
+try:
+    # Do imports
+    from .jadepy.jade import JadeAPI
+    from serial.tools import list_ports
+except ImportError as e:
+    _logger.exception('error importing Jade plugin deps')
+
+# Ignore -beta and -rc etc labels
+def _versiontuple(v):
+    return tuple(map(int, (v.split('-')[0].split('.'))))
+
+def _is_multisig(wallet):
+    return type(wallet) is Multisig_Wallet
+
+# Ensure a multisig wallet is registered on Jade hw.
+# Derives and returns the deterministic name for that multisig registration
+def _register_multisig_wallet(wallet, keystore, address):
+    wallet_fingerprint_hash = sha256(wallet.get_fingerprint())
+    multisig_name = 'ele' + wallet_fingerprint_hash.hex()[:12]
+
+    # Collect all the signer data in case we need to register the
+    # multisig wallet on the Jade hw - NOTE: re-register is a no-op.
+    signers = []
+    for kstore in wallet.get_keystores():
+        fingerprint = kstore.get_root_fingerprint()
+        bip32_path_prefix = kstore.get_derivation_prefix()
+        derivation_path = bip32.convert_bip32_path_to_list_of_uint32(bip32_path_prefix)
+
+        # Jade only understands standard xtypes, so convert here
+        node = bip32.BIP32Node.from_xkey(kstore.xpub)
+        standard_xpub = node._replace(xtype='standard').to_xkey()
+
+        signers.append({'fingerprint': bytes.fromhex(fingerprint),
+                        'derivation': derivation_path,
+                        'xpub': standard_xpub,
+                        'path': []})
+
+    # Check multisig is registered - re-registering is a no-op
+    # NOTE: electrum multisigs appear to always be sorted-multisig
+    txin_type = wallet.get_txin_type(address)
+    keystore.register_multisig(multisig_name, txin_type, True, wallet.m, signers)
+
+    # Return the name used to register the wallet
+    return multisig_name
+
+# Helper to adapt Jade's http call/data to Network.send_http_on_proxy()
+def _http_request(params):
+    # Use the first non-onion url
+    url = [url for url in params['urls'] if not url.endswith('.onion')][0]
+    method = params['method'].lower()
+    json_payload = params.get('data')
+    json_response = Network.send_http_on_proxy(method, url, json=json_payload)
+    return {'body': json.loads(json_response)}
```

Comment:
This looks good.

Could we also disable the fallback default `_http_request` implementation in jadepy, for sanity?
I don't really like this API where the http func has to be passed explicitly with every function call otherwise there is a silent fallback :P It has good usability but I would like to err on the explicit-failure-side of things here.

If we wanted to monkey-patch it from here, we could do e.g.
```python
try:
    from .jadepy import jade
except ImportError:
    pass
else:
    if hasattr(jade, "_http_request"):
        del jade._http_request
    del jade
```
But I guess it might be useful even for upstream to have some global flag in `jadepy` we can set to disable the default impl.

Reply:
ok for now we can just comment the code entirely or delete the function as in your patch.
In parallel, I'll work the suggested improvements into the jade repo, and at some point we can refresh the jadepy here from a later tag/release of Jade.

Reply:
Sure. Thanks.